File: simulator/tests/src/simulator_benchmark_test.cpp
#include <gtest/gtest.h>

#include <simulator/algorithms_creator.hpp>
#include <simulator/argument_parser.hpp>
#include <simulator/simulator.hpp>
#include <sstream>

class SimulatorBenchmarkTest : public ::testing::Test {
protected:
    void SetUp() override {
        algorithms_ =
            AlgorithmsCreator::CreateAlgorithms({"brute-force", "KMP", "Rabin-Karp", "Z-algorithm",
                                                 "Aho-Corasick", "suffix-tree", "boyer-moore"});
        algorithms_move_ =
            AlgorithmsCreator::CreateAlgorithms({"brute-force", "KMP", "Rabin-Karp", "Z-algorithm",
                                                 "Aho-Corasick", "suffix-tree", "boyer-moore"});
    }

    void VerifyOutput(const std::string& output, const std::string& algorithm_name,
                      bool expected_found) {
        std::istringstream iss(output);
        std::string line;
        while (std::getline(iss, line)) {
            if (line.find("Algorithm: " + algorithm_name) != std::string::npos) {
                std::getline(iss, line);  // Read Duration line
                std::getline(iss, line);  // Read Status line
                std::string expected_status =
                    "Status: Substring" + std::string(expected_found ? " " : " not ") + "found";
                ASSERT_EQ(line, expected_status);
                return;
            }
        }
        FAIL() << "No output for algorithm: " << algorithm_name;
    }

    ArgumentParser parser_{0, nullptr};  // Stub for ReadFileContent usage
    AlgorithmList algorithms_, algorithms_move_;
};

TEST_F(SimulatorBenchmarkTest, Text500k_Substring63_NotFound) {
    algorithms_ = AlgorithmsCreator::CreateAlgorithms({});
    algorithms_move_ = AlgorithmsCreator::CreateAlgorithms({});
    std::string text = parser_.ReadFileContent(
        "../../../simulator/tests/resources/test-text-500k-substring-63-not-found/text.txt");
    std::string substring = parser_.ReadFileContent(
        "../../../simulator/tests/resources/test-text-500k-substring-63-not-found/substring.txt");

    Simulator simulator(std::move(algorithms_move_));

    std::ostringstream output;
    simulator.Run(text, substring, output);

    for (auto& algorithm : algorithms_) {
        VerifyOutput(output.str(), algorithm->GetName(), false);
    }
}

TEST_F(SimulatorBenchmarkTest, Text500k_Substring63_Found) {
    algorithms_ = AlgorithmsCreator::CreateAlgorithms({});
    algorithms_move_ = AlgorithmsCreator::CreateAlgorithms({});
    std::string text = parser_.ReadFileContent(
        "../../../simulator/tests/resources/test-text-500k-substring-63-found/text.txt");
    std::string substring = parser_.ReadFileContent(
        "../../../simulator/tests/resources/test-text-500k-substring-63-found/substring.txt");

    Simulator simulator(std::move(algorithms_move_));

    std::ostringstream output;
    simulator.Run(text, substring, output);

    for (auto& algorithm : algorithms_) {
        VerifyOutput(output.str(), algorithm->GetName(), true);
    }
}

TEST_F(SimulatorBenchmarkTest, Text500k_Substring1k_NotFound) {
    std::string text = parser_.ReadFileContent(
        "../../../simulator/tests/resources/test-text-500k-substring-1k-not-found/text.txt");
    std::string substring = parser_.ReadFileContent(
        "../../../simulator/tests/resources/test-text-500k-substring-1k-not-found/substring.txt");

    Simulator simulator(std::move(algorithms_move_));

    std::ostringstream output;
    simulator.Run(text, substring, output);

    for (auto& algorithm : algorithms_) {
        VerifyOutput(output.str(), algorithm->GetName(), false);
    }
}

TEST_F(SimulatorBenchmarkTest, Text500k_Substring1k_Found) {
    std::string text = parser_.ReadFileContent(
        "../../../simulator/tests/resources/test-text-500k-substring-1k-found/text.txt");
    std::string substring = parser_.ReadFileContent(
        "../../../simulator/tests/resources/test-text-500k-substring-1k-found/substring.txt");

    Simulator simulator(std::move(algorithms_move_));

    std::ostringstream output;
    simulator.Run(text, substring, output);

    for (auto& algorithm : algorithms_) {
        VerifyOutput(output.str(), algorithm->GetName(), true);
    }
}

TEST_F(SimulatorBenchmarkTest, Text1M_Substring10k_NotFound) {
    std::string text = parser_.ReadFileContent(
        "../../../simulator/tests/resources/test-text-1m-substring-10k-not-found/text.txt");
    std::string substring = parser_.ReadFileContent(
        "../../../simulator/tests/resources/test-text-1m-substring-10k-not-found/substring.txt");

    Simulator simulator(std::move(algorithms_move_));

    std::ostringstream output;
    simulator.Run(text, substring, output);

    for (auto& algorithm : algorithms_) {
        VerifyOutput(output.str(), algorithm->GetName(), false);
    }
}

TEST_F(SimulatorBenchmarkTest, Text1M_Substring10k_Found) {
    std::string text = parser_.ReadFileContent(
        "../../../simulator/tests/resources/test-text-1m-substring-10k-found/text.txt");
    std::string substring = parser_.ReadFileContent(
        "../../../simulator/tests/resources/test-text-1m-substring-10k-found/substring.txt");

    Simulator simulator(std::move(algorithms_move_));

    std::ostringstream output;
    simulator.Run(text, substring, output);

    for (auto& algorithm : algorithms_) {
        VerifyOutput(output.str(), algorithm->GetName(), true);
    }
}

TEST_F(SimulatorBenchmarkTest, Text1M_Substring10k_NotFound_AntiBruteForce) {
    std::string text = parser_.ReadFileContent(
        "../../../simulator/tests/resources/test-text-1m-substring-10k-not-found-anti-bf/text.txt");
    std::string substring = parser_.ReadFileContent(
        "../../../simulator/tests/resources/test-text-1m-substring-10k-not-found-anti-bf/"
        "substring.txt");

    Simulator simulator(std::move(algorithms_move_));

    std::ostringstream output;
    simulator.Run(text, substring, output);

    for (auto& algorithm : algorithms_) {
        VerifyOutput(output.str(), algorithm->GetName(), false);
    }
}

File: simulator/tests/src/simulator_benchmark_test.py
import glob
import os
import platform
import psutil
import subprocess
import time

ALGORITHMS = ["KMP", "brute-force", "Rabin-Karp", "Z-algorithm", "shift-or", "Aho-Corasick", "suffix-tree", "boyer-moore"]


def run_simulator(algorithm_name, text_path, substring_path):
    simulator_base = ["simulator", "bin", "run-simulator"]

    if platform.system() == "Windows":
        simulator_base = ["simulator", "bin", "Release", "run-simulator.exe"]

    simulator_path = os.path.join(*simulator_base)

    command = [simulator_path, "--algorithm", algorithm_name, "-T", text_path, "-S", substring_path]

    if not os.path.exists(simulator_path):
        print(f"Simulator executable not found: {simulator_path}")
        return

    try:
        process = subprocess.Popen(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE)

        memory_used = 0

        # Monitor memory usage
        while process.poll() is None:
            mem_info = psutil.Process(process.pid).memory_info()
            current_memory = mem_info.rss / (1024 * 1024)  # Convert to MB
            if current_memory > memory_used:
                memory_used = current_memory
            time.sleep(0.01)

        stdout, stderr = process.communicate()

        if stderr:
            print(f"Error while executing algorithm {algorithm_name}: {stderr.decode()}")
            return

        print(f"Results for algorithm {algorithm_name}:\n{stdout.decode()}")
        print(f"Peak memory usage: {memory_used:.2f} MB\n-----------------------------------\n\n")

    except Exception as e:
        print(f"An error occurred while running algorithm {algorithm_name}: {e}")


def main():
    test_base_path = os.path.join("simulator", "tests", "resources")
    test_directories = [d for d in glob.glob(os.path.join(test_base_path, "test-*")) if os.path.isdir(d)]

    for test_dir in test_directories:
        text_path = os.path.join(test_dir, "text.txt")
        substring_path = os.path.join(test_dir, "substring.txt")

        if os.path.exists(text_path) and os.path.exists(substring_path):
            for alg in ALGORITHMS:
                print(f"Running simulator for algorithm: {alg}, directory: {test_dir}")
                run_simulator(alg, text_path, substring_path)
                time.sleep(1)
        else:
            print(f"text.txt or substring.txt not found in directory {test_dir}")


if __name__ == "__main__":
    main()
